emma-language/src/main/scala/org/emmalanguage/compiler/lang/comprehension/Combination.scala

```diff
@@ -540,88 +448,55 @@ private[comprehension] trait Combination extends Common {
-
-          val kyArg = api.TermSym.free(api.TermName.fresh("kyArg"), Core.bagElemTpe(yExpr))
-          val kyBdy = api.TopDown.transform { case core.Ref(`y`) => core.Ref(kyArg) }(
-            api.Tree.refreshAll(maybeAddCast(kyBdy0))).tree
-          val (kyRef, kyVal) = valDefAndRef("ky", core.Lambda(kyArg)(kyBdy))
-
-          val (irRef, irVal) = valDefAndRef("ir", Combinators.EquiJoin(kxRef, kyRef)(xExpr, yExpr))
-
-          //@formatter:off
-          val jTpe = Core.bagElemTpe(irRef)
-          assert(jTpe.typeConstructor == api.Sym.tuple(2).toTypeConstructor)
-          val jSym = api.TermSym.free(api.TermName.fresh("j"), jTpe)
-          val jRef = core.Ref(jSym)
-          val jx   = core.DefCall(Some(jRef))(tuple2_1)()
-          val jy   = core.DefCall(Some(jRef))(tuple2_2)()
-
-          val replaceXYandANF =
-            api.Tree.replace(core.Ref(x), jx) andThen
-            api.Tree.replace(core.Ref(y), jy) andThen
-            Core.anf
-
-          val qsReplFunc = this.qsReplFunc(replaceXYandANF)
-          val qs3Repl = qs3 map qsReplFunc
-          val qs4Repl = qs4 map qsReplFunc
-          val hdRepl  = cs.Head(Comprehension.asLet(replaceXYandANF(cs.Head.unapply(hd).get)))
-
-          cs.Comprehension(
-            qs1 ++
-            Seq(cs.Generator(jSym, core.Let(
-              xVals ++ yVals :+ kxVal :+ kyVal :+ irVal: _*
-            )()(irRef))) ++
-            qs3Repl ++ qs4Repl,
-            hdRepl
-          )
-          //@formatter:on
-
-        }).headOption
+    val MatchEquiJoin: Rule = {
+      case (owner, comp @ cs.Comprehension(qs, hd)) => (for {
+        // (Seq() is to disallow control flow)
+        xGen @ cs.Generator(x, core.Let(xVals, Seq(), xExpr)) <- qs.view
+        (qs1, qs234) = splitAt[u.Tree](xGen)(qs)
+        yGen @ cs.Generator(y, core.Let(yVals, Seq(), yExpr)) <- qs234.view
+        (qs2, qs34) = splitAt[u.Tree](yGen)(qs234)
+        if qs2.isEmpty && !api.Tree.refs(yGen).contains(x)
+        grd @ cs.Guard(core.Let(grdVals, Seq(), core.Ref(jcr))) <- qs34.view
+        if (api.Tree.refs(grd) intersect gens(qs)) == Set(x, y)
+        jcrVal @ core.ValDef(`jcr`, core.DefCall(Some(eqLhs), eq, _, Seq(eqRhs)), _) <- grdVals
+        if eq.name.toString == "$eq$eq"
+        joinVals = grdVals.filter(_ != jcrVal)
+        (kxExpr, kyExpr) <- Seq((eqLhs, eqRhs), (eqRhs, eqLhs))
+        kxBody = Core.dce(core.Let(joinVals: _*)()(kxExpr))
+        if !api.Tree.refs(kxBody).contains(y)
+        kyBody = Core.dce(core.Let(joinVals: _*)()(kyExpr))
+        if !api.Tree.refs(kyBody).contains(x)
+        (qs3, qs4) = splitAt[u.Tree](grd)(qs34)
+      } yield {
+        // It can happen that the key functions have differing return types (e.g. Long and Int).
+        // In this case, we add casts before the return expr of the key functions to the weakLub.
+        def maybeCast(tree: u.Tree) = if (kxExpr.tpe =:= kyExpr.tpe) tree else tree match {
```

Comment:
Whoops, we should use `=:=` to compare types, `==` unfortunately means reference equality.